### DISCUSSION TITLE
DKIM succeeds... and fails! - Partially RESOLVED

### DISCUSSION THREAD

**Discussion by eric-moreau** [PHPMailer/PHPMailer#2829] [Q&A] [answered]
I'm using PHPMailer to send email along with a DKIM signature.
Something like this:
```
$mail->DKIM_domain = "domain.com";                     // domain goes here
$mail->DKIM_private = "/home/...";                     // private key path goes here
$mail->DKIM_selector = "someselector";                 // key selector goes here
$mail->DKIM_passphrase = "somepwd";                    // key password goes here
$mail->DKIM_identity = $mail->From;                    // as recommended
```

And I do:
```
$mail->isHTML(true);
$mail->Body = $html;
```

Then DKIM signature is fine (I get a "pass").

But when instead I do:
```
$mail->isHTML(true);
$mail->Body = $html;
$mail->AltBody = $text;
```
Or also:
```
$mail->msgHTML($html);
```

Then DKIM signature fails to verify upon reception (I get an error when reading the email headers: "body was altered").

Have I gone into a common pitfall?
Shall I submit some more detailed code?

Thanks !

PS: this seems to be a more general issue: with AltBody, when adding attachments, etc.

**Comment by eric-moreau** [comment, marked as answer]
@Synchro just released version 6.7.
It fixes a couple problems... but not all.
I've made tests with Gmail and Office365.

**Fixed:**
alternate messages:
- Body="some HTML"; AltBody="some text";
as well as:
- MsgHTML("some HTML");

**Not yet fixed:**
- inline images made with: MsgHTML("some HTML", "some path");
- attachments made with: AddAttachment("some path");
Both are still reported with an incorrect hash upon reception.

**Still ok...**
Also, note that fortunately, plain messages (text or HTML only messages, like Body = ...) as still signed correctly, as this was already the case before version 6.7.

**Reply by Synchro** [reply]
I know there are some DKIM issues, but I also know they are extremely difficult to fix due to PHPMailer's legacy structure. DKIM is also not nice to work with as automated testing is difficult, and you essentially only ever get a pass/fail with no other clues about what the issue might be. There are numerous threads on DKIM on here such as #1655, #1965. Help appreciated!

**Reply by eric-moreau** [reply]
I see you've also worked on DKIM Validator. It doesn't help much I guess... as I believe the best thing would rather be a "step by step" signing (not verifying) code to challenge PHPMailer's code. I assume remaining issues are kind of related with MIME embeded binaries (inline images or attachments which are the same stuff in MIME).

**Reply by Synchro** [reply]
Yes I did a lot of work on that - there's not really anything left of the original in my branch! One thing I realised in there is that PHPMailer could really do with a proper abstraction of MIME components, and sane ways of handling headers. A key thing about DKIM is that signing is essentially the same thing as validating, so it would make sense to have abstract headers and bodies that can work in both scenarios. I did however grind to a  a halt on designing a structure that would work well there. DKIM is *really* picky, but the RFC is not an especially easy read, particularly when it comes to what order to assemble headers in and how you handle multiple DKIM signatures (because a second one can sign a first one, etc). I did get some help from the guy who wrote nodemailer about that, but never managed to see it through.
